### PR TITLE
Update object change context manager to be aware of the change source

### PR DESCRIPTION
**Pull request by gsnider2195**
# Closes: #1876
# What's Changed

- Add new fields to ObjectChange model
  - change_context
  - change_context_detail
- Add `ObjectChangeEventContextChoices` to `nautobot.extras.choices` for `ObjectChange.change_context` field choices
- Update `objectchange` template to add `change_context` and `change_context_detail` to the object detail view
- Add new `ChangeContext` class to be used as input to the `change_logging` context manager. This class stores the user, change context, change context detail and unique id for the change
- Add `ChangeContext` convenience subclasses for:
  - `JobChangeContext`
  - `WebChangeContext`
  - `ORMChangeContext`
- Update `web_request_context` context manager to use `ORMChangeContext`
- Update `run_job` to use `JobChangeContext`
- Update `handle_changed_object` and `handle_deleted_object` signals to use `ChangeContext` instead of a `WSGIRequest`

image
image

# TODO

- [x] Explanation of Change(s)
- [x] Attached Screenshots, Payload Example
- [x] Unit, Integration Tests
- [x] Documentation Updates (when adding/changing features)
- [x] Example Plugin Updates (when adding/changing features)
- [x] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/extras/models/change_logging.py [hunk 1]
```diff
@@ -67,6 +67,8 @@ class ObjectChange(BaseModel):
     changed_object_type = models.ForeignKey(to=ContentType, on_delete=models.PROTECT, related_name="+")
     changed_object_id = models.UUIDField(db_index=True)
     changed_object = GenericForeignKey(ct_field="changed_object_type", fk_field="changed_object_id")
+    change_context = models.CharField(max_length=50, choices=ObjectChangeEventContextChoices, blank=True)
+    change_context_detail = models.CharField(max_length=100, blank=True)
```
Comment (gsnider2195) [inline nautobot/extras/models/change_logging.py:71]: Should these be set to `editable=False` and added to `csv_headers`?

File: nautobot/extras/jobs.py [hunk 2]
```diff
@@ -1180,7 +1178,8 @@ def _run_job():
     # Execute the job. If commit == True, wrap it with the change_logging context manager to ensure we
     # process change logs, webhooks, etc.
     if commit:
-        with change_logging(request):
+        change_context = JobChangeContext(request.user, id=request.id, context_detail=job.name)
```
Comment (glennmatthews) [inline nautobot/extras/jobs.py:1181]: Wonder if a better `context_detail` value here would be `job_model.slug`?

File: nautobot/extras/models/change_logging.py (hunk 1 above)
Comment (glennmatthews) [inline nautobot/extras/models/change_logging.py:71]: `blank=True` is just here for pre-existing records only, right? Wonder if it'd be better to disallow blank values and either populate existing records as `web` or alternatively add an `unknown` value that can be used for those?

File: nautobot/extras/templates/extras/objectchange.html [hunk 3]
```diff
@@ -81,6 +81,18 @@
...
+                    <tr>
+                        <td>Change Context</td>
+                        <td>
+                            {{ object.get_change_context_display }}
+                        </td>
+                    </tr>
+                    <tr>
+                        <td>Change Context Detail</td>
+                        <td>
+                            {{ object.change_context_detail }}
```
Comment (glennmatthews) [inline nautobot/extras/templates/extras/objectchange.html:93]:
```suggestion
                            {{ object.change_context_detail | placeholder }}
```

File: nautobot/core/middleware.py [hunk 4]
```diff
@@ -1,13 +1,13 @@
...
 from django.contrib.auth.middleware import RemoteUserMiddleware as RemoteUserMiddleware_
 from django.db import ProgrammingError
 from django.http import Http404
+from django.urls import resolve
+from django.urls.exceptions import Resolver404
 from django.utils.deprecation import MiddlewareMixin
 
 from nautobot.core.views import server_error
-from nautobot.extras.context_managers import change_logging
+from nautobot.extras.context_managers import WebChangeContext, change_logging
```
Comment (timizuoebideri1) [inline nautobot/core/middleware.py:10]:
```suggestion
from nautobot.extras.context_managers import change_logging, WebChangeContext
```

**Review by glennmatthews — COMMENTED** [review]
This looks really promising! Nice work.

File: nautobot/extras/models/change_logging.py (hunk 1 above)
Comment (glennmatthews) [reply nautobot/extras/models/change_logging.py:71]: Good thought on both. Maybe also `db_index=True` on `change_context` at least?

File: nautobot/extras/datasources/registry.py [hunk 5]
```diff
@@ -1,7 +1,7 @@
 """Registry-related APIs for datasources."""
 
 from nautobot.extras.choices import JobResultStatusChoices, LogLevelChoices
-from nautobot.extras.context_managers import change_logging
+from nautobot.extras.context_managers import JobChangeContext, change_logging
```
Comment (timizuoebideri1) [inline nautobot/extras/datasources/registry.py:4]:
```suggestion
from nautobot.extras.context_managers import change_logging, JobChangeContext
```

File: nautobot/extras/models/change_logging.py [hunk 6]
```diff
@@ -93,6 +100,8 @@ class ObjectChange(BaseModel):
         "related_object_id",
         "object_repr",
         "object_data",
+        "change_context",
+        "change_context_detail",
```
Comment (glennmatthews) [inline nautobot/extras/models/change_logging.py:104]: Be sure to update `to_csv()` accordingly as well!

**Review by glennmatthews — APPROVED** [review]

**Pull request merged**